docs/plugins/first_plugin.md

```diff
@@ -0,0 +1,293 @@
+3. That our plugin contributes a [**widget**](./contributions.html#contributions-widgets),
+   and that we'd like napari to **autogenerate** the widget from the command signature
+   (so we don't need to deal with any GUI code).
+
+Add the following text to `napari.yaml`:
+
+```yaml
+name: napari-hello
+contributions:
+  commands:
+    - id: napari-hello.say_hi
+      title: Say hello, world!
+      python_name: napari_hello:show_hello_message
+  widgets:
+    - command: napari-hello.say_hi  # note, same as command.id above
+      display_name: Hello World
+      autogenerate: true
+```
+
+```{tip}
+We could have put the `show_hello_message` function anywhere, it didn't
+need to go in the top `__init__.py` file. Just make sure that the `python_name`
+of the corresponding `command` in the manifest points to the correct
+`<module_name>:<function_name>`.
+```
+
+### Update `setup.cfg`
+
+Lastly, let's make a few small changes to `setup.cfg`.
+
+1. Because we are directly using the `napari.utils.notifications` API in our,
+   `show_hello_message` function, we need to add `napari` to our package
+   **`install_requires`**. (You should add *all* of your package dependencies
+   here.  Assume nothing about your user's environment!  Not even napari.)
+
+2. We need to instruct setuptools to *include* that `napari.yaml` file
+   when it bundles our package for distribution, by adding it to the
+   **`[options.package_data]`**
+
+3. In order for napari to find our plugin when it's installed in
+   the environment, we need to add a `napari.manifest` entry to our
+   **`[options.entry_points]`** that points to the `napari.yaml`
+   file we added to the `napari_hello` module.
+
+    ```{tip}
+    Entry points are a standard Python mechanism for an installed distribution to
+    advertise components it provides to be discovered and used by other code.
+
+    See the [Entry points specification][entry_points] for details.
+    ```
+
+With the above changes, your final `setup.cfg` file should look like this:
+
+```ini
+[metadata]
+name = napari-hello
+version = 0.0.1
+classifiers =
+    Framework :: napari
+
+[options]
+packages = find:
+install_requires =
+    napari
+
+[options.package_data]
+napari-hello = napari.yaml
+
+[options.entry_points]
+napari.manifest =
+    napari-hello = napari_hello:napari.yaml
+```
+
+## 4. Install your plugin and try it out!
+
+With that, we're ready to go.  Let's install our package in the environment
+and then run napari.  Note: we're using `pip install -e .` here to install our
+package (in the current working directory) in ["editable" mode][editable_mode].
+This means that changes we make to our package during development will be
+detected when we re-run napari, without having to run `pip install` again.
+
+```sh
+pip install -e .
+napari
+```
+
+Once napari starts, select `napari-hello: Hello World` from the
+`Plugins` menu, then click the `Run` button to see the message.
+
+![hello-example](../images/hello.png)
```

Comment:
I think this could point to a url not to a file added to repo right?

Reply:
Just a reminder about this: removing the png in a commit is not enough, the commit where it was added needs to be removed from history and the branch force pushed!

Reply:
i thought if we squash merge then it's fine

Reply:
I forgot about the squash... I'm not sure actually, you might be right!